### PR TITLE
modified the possible list for input polarizations

### PR DESCRIPTION
**Pull request by oberonia78**
This PR adds the capability to use `['dual-pol'], ['single-pol'], ['cross-pol']` for the polarizations in runconfig. 
When these polarizations are provided, the algorithm will search for the corresponding polarizations within the input RTC files and utilize them for subsequent processing. This enhancement adds more flexibility in polarization selection for the algorithm.

### CODE REVIEW

File: src/dswx_sar/dswx_runconfig.py [hunk 1]
```diff
@@ -289,15 +347,22 @@ def load_from_yaml(cls, yaml_path: str, workflow_name: str, args):
...
-            algorithm_cfg['runconfig']['processing']['polarizations'])
-        
+            requested_pol, input_dir_list)
+
+        # update the polarizations
         algorithm_cfg['runconfig']['processing']['polarizations'] = pol_list
-        algorithm_cfg['runconfig']['processing']['copol'] = co_pol[0]
-        algorithm_cfg['runconfig']['processing']['crosspol'] = cross_pol[0]
+        algorithm_cfg['runconfig']['processing']['copol'] = co_pol[0] if co_pol else None
+        algorithm_cfg['runconfig']['processing']['crosspol'] = cross_pol[0] if cross_pol else None
```
Comment (seongsujeong) [inline src/dswx_sar/dswx_runconfig.py:362]: What is going to happen if there is a two different dual-pols in the burst RTC list?
i.e. `['VV', 'VH']` and `['HH', 'HV']`?

**Review by seongsujeong — COMMENTED** [review]
LGTM overall. My comments are as below.

File: src/dswx_sar/detect_inundated_vegetation.py
Comment (seongsujeong) [inline src/dswx_sar/detect_inundated_vegetation.py:1]: Could not leave the comment directly on the line, but it looks like the [line 24-25](
https://github.com/oberonia78/DSWX-SAR/blob/9eb34623c62ed62897140e6748e91fb5df520127/src/dswx_sar/detect_inundated_vegetation.py#L24-L25) and [line 28-29](https://github.com/oberonia78/DSWX-SAR/blob/9eb34623c62ed62897140e6748e91fb5df520127/src/dswx_sar/detect_inundated_vegetation.py#L28-L29) are duplicated. Please check.

File: src/dswx_sar/dswx_runconfig.py [hunk 2]
```diff
@@ -15,6 +15,19 @@
 
 WORKFLOW_SCRIPTS_DIR = os.path.dirname(dswx_sar.__file__)
 
+# Potential polarization scenarios for DSWx-S1
+# NOTE: DO NOT CHANGE ORDER of the dictionary.
```
Comment (seongsujeong) [inline src/dswx_sar/dswx_runconfig.py:19]:
```suggestion
# NOTE: DO NOT CHANGE THE ORDER of the items in the dictionary below.
```

**Review by seongsujeong — APPROVED** [review]
LGTM. Pending couple of minor comment and suggestion.

File: src/dswx_sar/dswx_runconfig.py (hunk 1 above)
Comment (oberonia78) [reply src/dswx_sar/dswx_runconfig.py]: I think this question is resolved now.  For example, if the bursts have four polarizations, they are going to be processed separately. 
In the above lines, `['VV']` and `['HH']` are assigned to the `copol`.

File: src/dswx_sar/save_mgrs_tiles.py [hunk 3]
```diff
@@ -155,37 +205,39 @@ def find_intersecting_burst_with_bbox(ref_bbox,
...
+                        transform_bounds(src.crs,
+                                        {'init': f'EPSG:{ref_epsg}'},
+                                        left,
+                                        bottom,
+                                        right,
+                                        top)
+                rtc_polygon = Polygon([(left, bottom),
+                                    (left, top),
+                                    (right, top),
+                                    (right, bottom)])
```
Comment (LiangJYu) [inline src/dswx_sar/save_mgrs_tiles.py:232]: nit - alignment
```suggestion
                                         {'init': f'EPSG:{ref_epsg}'},
                                         left,
                                         bottom,
                                         right,
                                         top)
                rtc_polygon = Polygon([(left, bottom),
                                       (left, top),
                                       (right, top),
                                       (right, bottom)])
```

File: src/dswx_sar/save_mgrs_tiles.py [hunk 4]
```diff
@@ -155,37 +205,39 @@ def find_intersecting_burst_with_bbox(ref_bbox,
...
-                                   (right, bottom)])
-
-        # Check if bursts intersect the reference polygon
-        if ref_polygon.intersects(rtc_polygon) or \
-           ref_polygon.overlaps(rtc_polygon):
-            overlapped_rtc_dir_list.append(input_dir)
+        copol_file_list = []
+        for copol in dswx_s1_pol_dict['CO_POL']:
+            copol_rtc_path = glob.glob(f'{input_dir}/*_{copol}*.tif')
+            copol_file_list.extend(copol_rtc_path)
```
Comment (LiangJYu) [inline src/dswx_sar/save_mgrs_tiles.py:211]: List comprehension version with more than1 glob call and filtering done in glob.
```suggestion
        copol_file_list = [f for pol in dswx_s1_pol_dict['CO_POL']
                           for f in glob.glob(f'{input_dir}/*_{copol}*.tif')]
```

List comprehension version with single glob call and filtering done to all tif files.
```suggestion
        copol_file_list = [f for f in glob.glob(f'{input_dir}/*.tif')
                           if any(pol in f
                                  for pol in dswx_s1_pol_dict['CO_POL'])]
```

File: src/dswx_sar/save_mgrs_tiles.py [hunk 5]
```diff
@@ -516,27 +569,49 @@ def run(cfg):
...
+                           for pol_cand_str in merge_pol_list]
+            if key == 'no_data_area':
+                extra_args = {'nodata_value': 1}
+            elif key == 'fuzzy_value':
+                extra_args = {'nodata_value': -1}
+            else:
+                extra_args = {'nodata_value': 0}
+            merge_pol_layers(list_layers,
+                         os.path.join(outputdir, file_path),
+                         **extra_args)
```
Comment (LiangJYu) [inline src/dswx_sar/save_mgrs_tiles.py:604]: nit - alignement
```suggestion
            merge_pol_layers(list_layers,
                             os.path.join(outputdir, file_path),
                             **extra_args)
```

File: src/dswx_sar/dswx_sar_util.py [hunk 6]
```diff
@@ -354,8 +355,8 @@ def get_invalid_area(geotiff_path,
     """
     image = read_geotiff(geotiff_path)
     if image.ndim == 3:
-        no_data_raster = np.isnan(
-            np.squeeze(image[0, :, :]))
+        no_data_raster = \
```
[the remark this answers is not in the document]
Comment (oberonia78) [reply src/dswx_sar/dswx_sar_util.py:358]: Yes. Your understanding is correct.

File: src/dswx_sar/refine_with_bimodality.py [hunk 7]
```diff
@@ -796,6 +796,9 @@ def remove_false_water_bimodality_parallel(water_mask,
     sizes = stats_water[1:, -1]
     bounding_boxes = stats_water[1:, :4]
 
+    bimodality_total = water_mask.copy()
```
[the remark this answers is not in the document]
Comment (oberonia78) [reply src/dswx_sar/refine_with_bimodality.py:799]: Yes. In the case that input polarization is not in the list, it returns input as output.
I added a comment on this.

File: src/dswx_sar/mosaic_rtc_burst.py [hunk 8]
```diff
@@ -21,6 +22,110 @@
 logger = logging.getLogger('dswx_s1')
 
 
+@dataclass
+class DSWX_geogrid:
```
[the remark this answers is not in the document]
Comment (oberonia78) [reply src/dswx_sar/mosaic_rtc_burst.py:26]: Yes. This class is moved to the new file `dswx_geogrid.py`

File: src/dswx_sar/refine_with_bimodality.py [hunk 9]
```diff
@@ -772,7 +774,7 @@ def remove_false_water_bimodality_parallel(water_mask,
     bimodality_total: numpy.ndarray
         An image indicating the bimodality values across the entire scene.
     """
-
+    print('pol_list', pol_list)
```
Comment (LiangJYu) [inline src/dswx_sar/refine_with_bimodality.py:777]: Log message or debug print statement here?

**Review by LiangJYu — APPROVED** [review]
LGTM - just one nit

**Pull request merged**